graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -445,71 +447,81 @@ def remove_backtrack_blocks_from_fold(folded_ir_blocks):
+
+    return new_traverse
+
+
+def convert_optional_traversals_to_compound_match_query(
+        match_query, optional_locations, location_to_optional):
     """Return 2^n distinct MatchQuery objects in a CompoundMatchQuery.
 
     Args:
         match_query: MatchQuery object potentially containing n `@optional` scopes
-            which expand vertex fields
+                     which expand vertex fields
+        optional_locations: List of locations with @optional that expand vertex fields within.
+        location_to_optional: Dict mappingall locations within optional scopes
+                              to the corresponding optional tag.
 
     Returns:
         CompoundMatchQuery object containing 2^n MatchQuery objects,
         one for each possible subset of the n optional edges being followed
     """
-    def filter_local_field_existence(field_name):
-        """Return an Expression that is True iff `field_name` does not exist."""
-        local_field = LocalField(field_name)
-        local_field_size = UnaryTransformation(u'size', local_field)
-        field_null_check = BinaryComposition(u'=', local_field, NullLiteral)
-        field_size_check = BinaryComposition(u'=', local_field_size, ZeroLiteral)
-        return BinaryComposition(u'||', field_null_check, field_size_check)
-
-    def is_optional_traverse(traverse):
-        """Return True if `traverse` contains any traversal within an @optional scope."""
-        return any(
-            isinstance(step.root_block, Traverse) and step.root_block.within_optional_scope
-            for step in traverse
-        )
-
-    def remove_optional_traverse(traverse):
-        """Return a new traversal, removing the optional traverse and all following steps."""
-        new_traverse = []
-        for step in traverse:
-            if not isinstance(step.root_block, Traverse) or not step.root_block.optional:
-                new_traverse.append(step)
-            else:
-                field_name = step.root_block.get_field_name()
-                new_predicate = filter_local_field_existence(field_name)
-                old_filter = new_traverse[-1].where_block
-                if old_filter:
-                    new_predicate = BinaryComposition(u'&&', old_filter.predicate, new_predicate)
-                new_traverse[-1] = new_traverse[-1]._replace(where_block=Filter(new_predicate))
-                return new_traverse
-        raise AssertionError(u"No optional traverse found in: {}".format(traverse))
-
-    def make_mandatory_traverse(traverse):
-        """Return a new traversal, without the optional condition."""
-        new_traverse = []
-        for step in traverse:
-            if not isinstance(step.root_block, Traverse) or not step.root_block.optional:
-                new_traverse.append(step)
+    optional_location_subsets = itertools.chain(
+        *[itertools.combinations(optional_locations, x)
+            for x in range(0, len(optional_locations)+1)]
```

Comment:
The formatting here is off. Also, there is a lot of complexity in this expression -- let's break it up a few lines with well-named variables.